design-documents/graph-ql/coverage/company.md

```diff
@@ -0,0 +1,309 @@
+    job_title: String! @doc(description: "Company User job title.")
+    telephone: String! @doc(description: "Company User phone number.")
+    role: CompanyRole! @doc(description: "Company User role data (includes permissions).")
+    team: CompanyTeam @doc(description: "Company User team data.")
+}
+
+type CompanyRoles @doc(description: "Output data schema for an object returned by a Company roles search query.") {
+    items: [CompanyRole] @doc(description: "A list of company roles that match the specified search criteria.")
+    total_count: Int @doc(description: "The total number of objects matching the specified filter.")
+    page_info: SearchResultPageInfo @doc(description: "Pagination meta data.")
+}
+
+type CompanyRole @doc(description: "Company role output data schema returned in response to a query by Role id.") {
+    id: ID! @doc(description: "Role id.")
+    name: String! @doc(description: "Role name.")
+    users_count: Int @doc(description: "Total number of Users with such Role within Company Hierarchy.")
+    permissions: [String] @doc(description: "A list of permission resources defined for a Role.")
+}
+
+type CompanyAclResource @doc(description: "Output data schema for an object with Role permission resource information.") {
+    id: ID! @doc(description: "ACL resource id.")
+    text: String! @doc(description: "ACL resource label.")
+    sortOrder: Int! @doc(description: "ACL resource sort order.")
+    children: [CompanyAclResource!] @doc(description: "An array of sub-resources.")
+}
+
+type CompanyUserEmailCheckResponse @doc(description: "Response object schema for a Company User email validation query.") {
+    status: Boolean! @doc(description: "True - email is valid, False - email is not valid")
+}
+
+type CompanyHierarchyOutput @doc(description: "Response object schema for a Company Hierarchy query.") {
+    structure: CompanyHierarchyElement @doc(description: "An array of Company structure elements.")
+    draggable: Boolean @doc(description: "Flag that defines whether Company Hierarchy can be changed by current User or not.")
+    max_nesting: Int @doc(description: "Indicator of maximun nesting of elements within a whole Company Hierarchy.")
+}
+
+type CompanyHierarchyElement @doc(description: "Company Hierarchy element output data schema.") {
+    id: ID! @doc(description: "Hierarchy element id.")
+    tree_id: ID! @doc(description: "The hierarchical id of the element within a structure. Used for changing element's position in hierarchy.")
+    type: String! @doc(description: "Hierarchy element type: 'customer' or a 'team'.")
+    text: String! @doc(description: "Hierarchy element name.")
+    description: String @doc(description: "Hierarchy element description.")
+    opened: Boolean! @doc(description: "Flag that defines whether Hierarchy element should be expanded or not.")
+    children: [CompanyHierarchyElement!] @doc(description: "An array of child elements.")
```

Comment:
I don't think this pattern of recursive `children` is going to scale very well for consumers.

If I want to go 4 levels deep from a client, my query has to be:

```graphql
query GetFourLevelsDeep {
	hierarchy {
        structure {
           type
           text
           children {
               type
               text
               children {
                   type
                   text
                   children {
                      type
                      text
                   }
               }
           }
        }
    }
}
```

This means that a client would have to dynamically generate their query with some form of string generation. [This is not great, as declaring queries statically has major benefits](https://blog.apollographql.com/5-benefits-of-static-graphql-queries-b7fa90b0b69a).

We've [already had this problem with the CMS schema](https://twitter.com/IvanChepurnyi/status/1202876134003740672).

A better alternative would be to return a flat list with `depth` keys, and take an input parameter specifying max depth. Then, a consumer can just switch their tree walking algorithm.

Reply:
We considered this approach, but decided to proceed with tree structure. We don't expect a lot of nesting here. Tree structure will simplify client side. There is "flat" alternative which allows to fetch all users without parent/child relations.